test/tools/cmap_spec_runner.ts

```diff
@@ -372,6 +390,102 @@ export async function runCmapTest(test: CmapTest, threadContext: ThreadContext)
     const actual = actualEvents.shift();
     const { type: eventType, ...eventPropsToCheck } = expected;
     expect(actual.constructor.name).to.equal(`${eventType}Event`);
-    compareInputToSpec(actual, eventPropsToCheck);
+    compareInputToSpec(actual, eventPropsToCheck, `failed while checking ${eventType} event`);
+  }
+}
+
+export type SkipDescription = {
+  description: string;
+  skipIfCondition: 'loadBalanced';
+  skipReason: string;
+};
+
+export function runCmapTestSuite(
+  tests: CmapTest[],
+  options?: { testsToSkip?: SkipDescription[]; injectPoolStats?: boolean }
+) {
+  for (const test of tests) {
+    describe(test.name, function () {
+      let hostAddress: HostAddress, threadContext: ThreadContext, client: MongoClient;
+
+      beforeEach(async function () {
+        let utilClient: MongoClient;
+
+        const skipDescription = options?.testsToSkip?.find(
+          ({ description }) => description === test.description
+        );
+        if (skipDescription) {
+          const matchesLoadBalanceSkip =
+            skipDescription.skipIfCondition === 'loadBalanced' && this.configuration.isLoadBalanced;
+          if (matchesLoadBalanceSkip) {
+            (this.currentTest as Mocha.Test & { skipReason?: string }).skipReason =
+              skipDescription.skipReason;
+            this.skip();
+          }
+        }
+
+        if (this.configuration.isLoadBalanced) {
+          // The util client can always point at the single mongos LB frontend.
+          utilClient = this.configuration.newClient(this.configuration.singleMongosLoadBalancerUri);
+        } else {
+          utilClient = this.configuration.newClient();
+        }
+
+        await utilClient.connect();
+
+        const allRequirements = test.runOn || [];
+
+        const someRequirementMet =
+          !allRequirements.length ||
+          (await isAnyRequirementSatisfied(this.currentTest.ctx, allRequirements, utilClient));
+
+        if (!someRequirementMet) {
+          await utilClient.close();
+          this.skip();
+          // NOTE: the rest of the code below won't execute after the skip is invoked
```

Comment:
I know TS warns you but another way we can help indicate this to readers is to `return this.skip()` but no need to change.

Reply:
I've considered it (when I first wrote this in the other PR), but I went against it because the return type for these functions is supposed to be `void`, so returning `this.skip()` would contradict the type, and placing a return after the skip highlights it as unreachable, hence the comment instead